### ISSUE TITLE
Toggle Superuser status on/off

### ISSUE THREAD

**Issue by djbrooke** [IQSS/dataverse#3612]
As a Superuser I want to add another Superuser so that I can have assistance in managing my dataverse installation.

This was an available feature in 3.x that we need to bring back in 4.x.

You should able to search for users by user name:

<img width="1134" alt="search for admin" src="https://cloud.githubusercontent.com/assets/7773392/22494520/2530c234-e806-11e6-9cef-91cd0dcf7f17.png">

... and then select a user to grant privileges to:

<img width="1102" alt="add and remove admin" src="https://cloud.githubusercontent.com/assets/7773392/22494529/346ab8d6-e806-11e6-9ed9-1b614b59836c.png">

**Comment by bencomp** [comment]
I thought there was an API endpoint that allowed superusers to "toggle" the superuser status of any user. I (still) think toggling instead of setting is not the right way to do it (see #2219), but (via the API) this is supported, AFAIK.

**Comment by pdurbin** [comment]
@bencomp you're right that there's an "admin" API endpoint for toggling the superuser boolean on and off. It's documented as "Toggles superuser mode on the AuthenticatedUser whose identifier (without the @ sign) is passed." at http://guides.dataverse.org/en/4.6/api/native-api.html#admin

For more context, here are the related issues: #3614, #3615, #3613.

**Comment by raprasad** [comment]
- business logic should be shared between this feature and API endpoint
   - or UI can use API endpoint

**Comment by raprasad** [comment]
- Consider extra security--e.g. re-enter password

**Comment by djbrooke** [comment]
- Should we ask the user to reaffirm password?
- Superuser is a status and not a role

**Comment by bencomp** [comment]
@djbrooke what is the difference between status and role? Being a superuser comes with power and superusers have a responsibility with regard to the system/service, unlike having (what I'd say is) a status like first-time user.

**Comment by pdurbin** [comment]
@bencomp hi! I wouldn't read too much into the word "status". It was just a word we were throwing around during a sprint planning meeting yesterday.

Here's the most official description we have about what a superuser is, in contrast to a user with an Admin role:

> The user who creates a dataverse is given the “Admin” role on that dataverse. The root dataverse is created automatically for you by the installer and the “Admin” is the superuser account (“dataverseAdmin”) we used in the Installation section to confirm that we can log in. These next steps of configuring the root dataverse require the “Admin” role on the root dataverse, but not the much more powerful superuser attribute. In short, users with the “Admin” role are subject to the permission system. A superuser, on the other hand, completely bypasses the permission system. You can give non-superusers the “Admin” role on the root dataverse if you’d like them to configure the root dataverse. -- http://guides.dataverse.org/en/4.6.1/installation/config.html#root-dataverse-configuration

As you can see, there I called being a superuser an "attribute". In practice, it's a boolean on the `authenticateduser` table, which you can see below and at http://phoenix.dataverse.org/schemaspy/latest/tables/authenticateduser.html

![screen shot 2017-06-01 at 6 38 04 am](https://cloud.githubusercontent.com/assets/21006/26676256/fde352bc-4694-11e7-9e19-8e2147e1a067.png)

The important thing is that someone with the `superuser` boolean set to `true` **completely bypasses the permission system**. They can do anything on the behalf of any user. Obviously, only a few trusted people should have this power! It's the equivalent of having "root" on a Unix system.

Yesterday I was starting to say that once you have the superuser boolean set to true, you don't need to have any roles in Dataverse but @scolapasta reminded me that notifications are tied to roles so a superuser might want to have roles on some dataverses or datasets to get notifications on various events, especially if that superuser is involved in curation of those datasets.

While I'm writing, I was just told at https://groups.google.com/d/msg/dataverse-community/tTb4uT7LfW4/D0tm-Q8EBAAJ that the "Dataverse Personas" doc at https://docs.google.com/document/d/1DeFRHFj41aF0K7ETOmuUtjNh3uc4q0VdsVphSP7WAqs/edit?usp=sharing I sent around was helpful. It might be worth taking a look.

**Comment by landreev** [comment]
This issue was discussed as the prime candidate to be merged with #3614. Indeed this one depends on #3614 100%. As it is about adding a very specific piece of functionality to a new page/view, that's currently being added as part of #3614. It was justified to open this issue as a separate one, because at that point it still wasn't decided where this "toggling" UI will be added/implemented. Since then however the decision was made to add it to the new dashboard users display. Hence the dependency on #3614. 

While it's still a separate issue, it appeared that it would make the most sense for me to check in the implementation in the 3614-user-list branch. An alternative would be to wait for #3614 to be completed and merged into develop; OR for me to make a pull request from a branch based on the current snapshot of #3614. The former is a realistic possibility, and something I can still do. This superuser toggling can be tested as is; but overall it could be confusing, as there are other things about this new dashboard that are still unfinished. So either way, it looks like we have to wait for the other, bigger issue to be done before this one can be moved into done as well.

**Comment by landreev** [comment]
I do feel like I should go ahead and change the /api/admin/superuser to *set* the superuser status, rather than *toggle* it; (with a PUT, instead of POST). 
I'm seeing that it's being used in the setup-all script; and in UtilIT.java, to facilitate restassured tests. So I would change it there too. 

I agree that "toggling" it, in the context of an API call just feels wrong. 

That said, if anyone wants to stop be from fixing it, on account of how it ain't broke, I'll listen.

**Comment by pdurbin** [comment]
@landreev please do not introduce a backward incompatible change to the API.

**Comment by landreev** [comment]
I would not, normally. But this particular API call is only being used, in real life, by the setup-all; and by restassured tests. I.e. always in the context of it happening in the same version of the tree. 
I'm totally ok with not touching it; just don't feel that backward compatibility is a particularly strong argument here.

**Comment by landreev** [comment]
(so, in other words, I'm not touching it)

**Comment by pdurbin** [comment]
@landreev heh. Thanks. I'm glad we're keeping the scope of this issue in check. Let's circle back to potentially changing the security around this API in the future, possibly when if and when we get around to looking at #2219.

**Comment by landreev** [comment]
I'd like somebody to review my UI solution for this; (in 3614-user-list branch). 
The QA, etc. will have to be done together with #3614, as this issue has no pull request of its own.

**Comment by landreev** [comment]
(OK, so @mheppler and I just had a micro review of this in the hallway, and Mike could tell right away he felt strongly about the need of a confirmation popup, before the status change is saved; so I'm going to go ahead and add it now; I left some stubs in the code to be able to do it quickly)

**Comment by landreev** [comment]
OK, now with a confirmation popup. 
@mheppler - please check how it looks, if anything needs to be made prettier.

**Comment by landreev** [comment]
Per  conversation with @mheppler, dragged it back into code review; to be reviewed together with the parent issue, #3614.

**Comment by landreev** [comment]
There were 2 code changes requested by @scolapasta: 

- Prevent the admin user currently viewing the dashboard from revoking the superuser status from him- or herself; this is done, and checked in. 

- Create commands for toggling the superuser status, to be used by the page and the api; this has been done - but not checked in, because I'm having some concerns now, whether that was worth the trouble. Gustavo, can we quickly review this tomorrow?

**Comment by mheppler** [comment]
Added the "Superuser" styling from the navbar to the Manage Users pg.

![screen shot 2017-06-28 at 12 30 23 pm](https://user-images.githubusercontent.com/687227/27648769-9c18d54c-5bfd-11e7-880e-fc5ff7c141bd.png)

**Comment by kcondon** [comment]
Works as advertised. Remaining tasks: sanity check with production data, potentially document api endpoints and try them out.
